Improve prek hook to check for proper imports in shared distributions

potiuk opened the pull request:
Since we are quite close to having a completed separation between
distributions, this PR will be rebased until we will fix all the
issues - we might want to extract things from this PR as we
rebase it and move the code between distributions.

The following checks were added:

* The shared distributions (neither in src nor tests) should never
  import `from airflow.`

* The distributions that are using shared code "airlfow-core",
  "task-sdk" should never import anything from either
  "airflow_shared" or from "_shared" import coming from another
  distribution.

* Neither "apache-airflow", "apache-airflow-task-sdk", "apache-airflow-core"
  should be added as dependency in a shared distribution.

* The "devel-common" should also not refer to airflow.* - it should only
  import from airflow_shared.*

<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->



<!-- Please keep an empty line above the dashes. -->
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [airflow-core/newsfragments](https://github.com/apache/airflow/tree/main/airflow-core/newsfragments).

potiuk committed:
Improve prek hook to check for proper imports in shared distributions

Since we are quite close to having a completed separation between
distributions, this PR will be rebased until we will fix all the
issues - we might want to extract things from this PR as we
rebase it and move the code between distributions.

The following checks were added:

* The shared distributions (neither in src nor tests) should never
  import `from airflow.`

* The distributions that are using shared code "airlfow-core",
  "task-sdk" should never import anything from either
  "airflow_shared" or from "_shared" import coming from another
  distribution.

* Neither "apache-airflow", "apache-airflow-task-sdk", "apache-airflow-core"
  should be added as dependency in a shared distribution.

* The "devel-common" should also not refer to airflow.* - it should only
  import from airflow_shared.*

jscheffl reviewed (approved):
Cool!

amoghrajesh commented on shared/module_loading/src/airflow_shared/module_loading/__init__.py:
```diff
@@ -67,3 +67,25 @@ def qualname(o: object | Callable) -> str:
     return pkgutil.iter_modules(ns.__path__, ns.__name__ + ".")
+
+
+def is_valid_dotpath(path: str) -> bool:
```
We already have defined `import_string` in sdk and core as well and it seems to be a fairly small function module to just have as duplicates in both vs having a shared library for it

potiuk replied:
But with this mechanism, I think it does not cost us much - almost nothing - to get those common methods "DRY" - with the shared distrubutions, we actually solved the issues that DRY code also has bad "coupling" side-effects. This mechhanism we have noww allows us to truly share code without any serious side effects.

I see no reason why we should **not** aim to be more DRY here :)

amoghrajesh replied:
I was trying not to have another "utils" -> it literally becomes a dumping ground very soon!

potiuk replied:
1) I am not so sure to be honest. We have almost completed the shared code extraction - there is not much left

2) I am not sure if it is a problem if we have more of those. The setup we have now is rally easy for sharing and it's super easy to "reuse" a code without any of the side effects.  I think it's **precisely** as I imagine sharing code even for small modules. No coupling, perfect IDE support, prek hook to make sure that imports are proper, building works out of the box, testing is isolated and does not depend on anything else (and uv sync takes care of it). Also each "module" has it's own dependencies that are individually automatically contributed to the user's dependencies when the users starts using particular module.

I think we now should see what else we can extract this way - rather than what to copy - especially if those are isolated, independent functtions.

This setup we have now has pretty much all DRY benefits and no DRY drawbacks (coupling) - I see no reason why "shared" shoud not have more of those isolated "util" things.

There is a **huge** problem of course if "util" is a **single bag of everything** and we should **definitely** not do it. I literally hate when you have it and when utils starts using each other in uncontrolled way (which happens very quickly when you have shared bag of utils). And when using one util means that you are signing up for all of them even if you do not want. But it's not our case on either front.

We have not "bag of everything" but several, "single purpose easily shareable and isolated modules" - which is very different than "bag of utils". mostly because those  separate shared modules will be independent from each other (mostly) and each of the "users" will only use those specific modules when needed - without necessarily importing all the different kinds of utils they do not need. Each of them has only dependencies it needs (if at all) and they are really nicely separated out from each other due to the way how workspace works and how we run tests for it. The cross-shared-modules-dependencies have no chance to slip through our harness of workspace / uv sync dependencies and import check.

This is **very** different than `bag of utils` IMHO.

potiuk replied:
I wonder if someone is able to say what problems such sharing might cause - even for small modules - with this setup. 

To be honest - I thought very hard and I can't think of any problem we can prevent by having "less" of those shared modules.

But maybe there is some I do not see ?

potiuk replied:
BTW. I can name it "shared/import-module" if `util` name is a problem :)

amoghrajesh replied:
My main concern around having too many shared libraries is _confusion_. It is very easy even for people using IDE support to end up importing `airflow._shared` in sdk code. Created unwanted and avoidable coupling.

There is also another risk (not specific to this case), lets take example of exceptions. If those are shared, `except MyException` would not catch sometimes just because the code that threw it was in `airflow.sdk._shared` but catching code is in `airflow._shared`.

To be clear, I do not have big objection, but thats my thinking in limiting shared libraries _only_ to whats really needed

potiuk replied:
I see the point. I don't think we should **actively** create many utils, but when we already have mostly copied code, I think it's definitely worth it - even if it's a small code. 

We have a lot of protection already that even if someone gets confused initially, either the pytest tests, or prek hook will fail if wrong import is used. For example if there will be a distribution when somene imports from `airflow._shared` - the pytests run in CI will fail. Whe you run `uv sync` in `task-sdk` code for example (once we complete the isolation) will **NOT** have `airflow._shared` in PYTHONPATH - because the distribution with that code will not be in `pyproject.toml` and:

```bash
cd task-sdk
uv run pytest
```

will fail. 

Plus we will add all the prek hooks and checks to provide very detailed explanation of why certain bad imports are wrong and how to fix them. Which is also a great self-learning exercise, because people will be explained or have to learn how things work as part of this.

I think it's worth it, especially that `shared` projects concept is here to stay and people will have to get used to it anyway - one more or less library following exactly the same pattern - and one that is easy usable - has at most opportunity to let people learn the patterns more quickly and easily.

amoghrajesh replied:
Yeah that's a valid point, I am OK with that :)

potiuk replied:
I will start extracting things out and fixing the test dependencies then :)

amoghrajesh commented on airflow-core/docs/authoring-and-scheduling/serializers.rst:
```diff
@@ -87,7 +87,7 @@ Registered
     from typing import TYPE_CHECKING
 
-    from airflow.utils.module_loading import qualname
+    from airflow._shared.module_loading import qualname
```
Let's not have shared paths in docs when possible. For this case, since we are going to move serde and serializers to task sdk, we can have it consume the `airflow.sdk` path instead by re-exporting it better

potiuk replied:
Yes. Same thought.

I was thinking that in airflow.sdk, the `_shared` should really be treated as `private` and if we want to expose something outside we should make a corresponding module that will be at `airflow.sdk.` level - just importing things from `._shared`  if we will think about making it available for Dag Authors. This way we can also only expose what we **really** think should be exposed rather than everything from the `shared` modules. WDYT?

amoghrajesh replied:
Precisely, take this for example: https://github.com/apache/airflow/blob/main/task-sdk/src/airflow/sdk/timezone.py,

We only expose what's needed through sdk really.

potiuk replied:
Fantastic!

amoghrajesh commented on airflow-core/src/airflow/serialization/serde.py:
```diff
@@ -31,10 +31,10 @@
 import attr
 
 import airflow.serialization.serializers
+from airflow._shared.module_loading import import_string, iter_namespace, qualname
```
Note to self: revisit this occurence as part of [Move over serde library to task sdk](https://github.com/apache/airflow/issues/58885)

amoghrajesh commented on airflow-core/src/airflow/settings.py:
```diff
@@ -636,6 +636,7 @@ def _configure_secrets_masker():
     core_masker.sensitive_variables_fields = list(sensitive_fields)
     core_masker.secret_mask_adapter = secret_mask_adapter
 
+    # TODO: this should be moved out when settings are moved to `shared`
```
I dont think I follow this one

potiuk replied:
We are importing "sdk" secrets masker here deliberately - and that should not happen eventually, because `airflow-core` should not have task-sdk installed at all. This import will not work  after that.

amoghrajesh commented on airflow-core/tests/unit/always/test_project_structure.py:
```diff
@@ -59,7 +59,6 @@ def test_providers_modules_should_have_tests(self):
         """
         # The test below had a but for quite a while and we missed a lot of modules to have tess
         # We should make sure that one goes to 0
-        # TODO(potiuk) - check if that test actually tests something
```
Hah!

potiuk replied:
It does :)

amoghrajesh commented on providers/apache/kafka/src/airflow/providers/apache/kafka/hooks/consume.py:
```diff
@@ -21,7 +21,7 @@
 
 from airflow.providers.apache.kafka.hooks.base import KafkaBaseHook
-from airflow.utils.module_loading import import_string
+from airflow.providers.common.compat.module_loading import import_string
```
Yay! These provider changes look very nice

amoghrajesh commented on scripts/ci/prek/check_shared_distributions_structure.py:
```diff
@@ -202,20 +372,217 @@ def check_shared_distribution(shared_path: Path) -> bool:
+    # Check devel-common distribution
+    if not check_devel_common_distribution():
+        all_ok = False
+
```
Changes to this file look good

amoghrajesh commented on shared/configuration/src/airflow_shared/configuration/parser.py:
```diff
@@ -1149,7 +1149,7 @@ def getimport(self, section: str, key: str, **kwargs) -> Any:
         try:
             # Import here to avoid circular dependency
-            from airflow.utils.module_loading import import_string
+            from ..module_loading import import_string
```
Nice

amoghrajesh commented on shared/module_loading/src/airflow_shared/module_loading/__init__.py:
```diff
@@ -14,21 +15,24 @@
-
     Raise ImportError if the import failed.
     """
+    # TODO: Add support for nested classes. Currently, it only works for top-level classes.
```
Do we need to? I dont see a use case for it

potiuk replied:
Nope. I just copied it from the original code.

amoghrajesh reviewed (approved):
Some observations and comments, looks good otherwise.

potiuk commented on task-sdk/src/airflow/sdk/module_loading.py:
```diff
@@ -1,24 +0,0 @@
-# Licensed to the Apache Software Foundation (ASF) under one
```
We do not need it for compatibility I guess ? This is not something that is part of the public `task-sdk` API?

potiuk replied:
Ah no... we do - common-compat uses it.

potiuk committed:
Skip TYPE_CHECKING imports and fix typos in shared distribution check

* Skip imports inside ``if TYPE_CHECKING:`` blocks so type-only imports
  in shared distributions (e.g. ``airflow.sdk.types`` references) no
  longer trigger false-positive violations.
* Fix files-pattern regex in the pre-commit hook: ``^shared/.*$||`` had
  a stray empty alternation, and ``.py`` / ``.toml`` dots were unescaped.
* Fix wrong ``airflow._shared.timezones`` import in task-sdk deadline
  deserialization (should be ``airflow.sdk._shared.timezones``).
* Fix typos in selective_checks.py hint message.
